base_layer/core/src/transaction_protocol_manager.rs

```diff
@@ -0,0 +1,1046 @@
+//! |             |       | AddChangeOutput    |             v
+//! |             +<------+                    |        +----+------+
+//! |             |                            |        |           |
+//! |             +-------+                    |        | Receiver  |
+//! |             |       | SetSenderNonce     |        | Partial   |
+//! |             +<------+                    |        | Signature |
+//! |             |                            |        |           |
+//! +-----+-------+                            |        +----+------+
+//!       |                                    |             |
+//!       | SetOffset                          |             | SetReceiverNonce
+//!       v                                    |             v
+//! +-----+-------+                            |        +----+------+
+//! |             |                            |        |           |
+//! | Receiver    |                            |        | Sender    |
+//! | Output      +----------------------------+        | Partial   |
+//! | Setup       | ConstructSenderPublicData           | Signing   |
+//! |             |                                     |           |
+//! +-----+-------+                                     +----+------+
+//!       |                                                 |
+//!       | AcceptReceiverPublicData<-----------------------+ ConstructReceiverPublicData
+//!       v
+//! +-----+-------+
+//! |             |
+//! | Sender      |
+//! | Partial     |
+//! | Signing     |
+//! |             |
+//! +-----+-------+
+//!       |
+//!       | SenderFinalize
+//!       v
+//! +-----+-------+
+//! |             |
+//! | Finalized   |       +---------+
+//! | Transaction +------>+ \o/ Fin |
+//! |             |       +---------+
+//! +-------------+
+
+use crate::{
+    transaction::{TransactionInput, TransactionOutput},
+    types::{BlindingFactor, SecretKey, Signature},
+};
+
+// use crate::types::PublicKey;
+use crate::{
+    transaction::{KernelFeatures, Transaction, TransactionBuilder, TransactionError, TransactionKernel},
+    types::{CommitmentFactory, PublicKey},
+};
+use crypto::{
+    challenge::Challenge,
+    commitment::HomomorphicCommitmentFactory,
+    common::Blake256,
+    keys::PublicKey as PublicKeyTrait,
+    signatures::SchnorrSignatureError,
+};
+use derive_error::Error;
+use tari_utilities::ByteArray;
+
+#[derive(Clone, Debug, PartialEq, Error)]
+pub enum TransactionProtocolError {
+    // The current state is not yet completed, cannot transition to next state
+    IncompleteStateError,
+    // Validation of the transaction being built failed, rectify issue and try again
+    ValidationError,
+    // Invalid state transition
+    InvalidTransitionError,
+    // Invalid state
+    InvalidStateError,
+    // An error occurred while performing a signature
+    #[error(no_from)]
+    SignatureError(SchnorrSignatureError),
```

Comment:
Perhaps `SigningError` to differentiate it from `InvalidSignatureError`?

Reply:
Sure, done